nopassword/forms.py

```diff
@@ -1,61 +1,133 @@
 # -*- coding: utf-8 -*-
 from django import forms
-from django.contrib.auth import authenticate
+from django.conf import settings
+from django.contrib.auth import authenticate, get_backends, get_user_model
+from django.contrib.sites.shortcuts import get_current_site
+from django.core.exceptions import ImproperlyConfigured
+from django.shortcuts import resolve_url
 from django.utils.translation import ugettext_lazy as _
 
-from .models import LoginCode
-from .utils import get_username_field
+from nopassword import models
 
 
-class AuthenticationForm(forms.Form):
-    """
-    Base class for authenticating users. Extend this to get a form that accepts
-    username logins.
-    """
-    username = forms.CharField(label=_("Username"), max_length=30)
-
+class LoginCodeRequestForm(forms.Form):
     error_messages = {
-        'invalid_login': _("Please enter a correct username. "
-                           "Note that it is case-sensitive."),
-        'no_cookies': _("Your Web browser doesn't appear to have cookies "
-                        "enabled. Cookies are required for logging in."),
+        'invalid_username': _(
+            "Please enter a correct %(username)s. "
+            "Note that it is case-sensitive."
+        ),
         'inactive': _("This account is inactive."),
     }
 
+    def __init__(self, *args, **kwargs):
+        super(LoginCodeRequestForm, self).__init__(*args, **kwargs)
+
+        self.username_field = get_user_model()._meta.get_field(get_user_model().USERNAME_FIELD)
+        self.fields['username'] = self.username_field.formfield()
+
+    def clean_username(self):
+        username = self.cleaned_data['username']
+
+        try:
+            user = get_user_model()._default_manager.get_by_natural_key(username)
+        except get_user_model().DoesNotExist:
+            raise forms.ValidationError(
+                self.error_messages['invalid_username'],
+                code='invalid_username',
+                params={'username': self.username_field.verbose_name},
+            )
+
+        if not user.is_active:
+            raise forms.ValidationError(
+                self.error_messages['inactive'],
+                code='inactive',
+            )
+
+        self.cleaned_data['login_code'] = models.LoginCode.create_code_for_user(user)
```

Comment:
Backward incompatible: Login code is not stored as `self.user_cache`. Variable name was misleading, as it is a login code. It is also easier to store in `cleaned_data`, because this way the rest view can access it as well (doesn't have access to the form).